File: 101.SymmetricTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    var isChildSymmetric = function(left, right) {
        if (!left && !right) {
            return true;
        } else if (!left || !right) {
            return false;
        } else if (left.val !== right.val) {
            return false;
        } else {
            return isChildSymmetric(left.left, right.right) && isChildSymmetric(left.right, right.left)
        }
    }
    if (!root || (!root.left && !root.right)) {
        return true
    }
    return isChildSymmetric(root.left, root.right);
};

var isSymmetric = function(root) {
    if (!root || (!root.left && !root.right)) {
        return true
    }
    const stack = [[root.left, root.right]];
    while(stack.length) {
        const temp = stack.shift();
        if (!temp[0] && !temp[1]) {
            continue;
        } else if (!temp[0] || !temp[1]) {
            return false;
        } else if (temp[0].val !== temp[1].val) {
            return false;
        } else {
            stack.push([temp[0].left, temp[1].right]);
            stack.push([temp[0].right, temp[1].left])
        }
    }
    return true;
};

File: Golang/Tree/101.symmetric-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	var recursive func(n1, n2 *TreeNode) bool
	recursive = func(n1, n2 *TreeNode) bool {
		if (n1 != nil && n2 == nil) || (n1 == nil && n2 != nil) {
			return false
		} else if n1 != nil && n2 != nil {
			if n1.Val == n2.Val {
				return recursive(n1.Left, n2.Right) && recursive(n1.Right, n2.Left)
			}
			return false
		}
		return true
	}
	return recursive(root.Left, root.Right)
}

func isSymmetricBFS(root *TreeNode) bool {
	if root != nil {
		stack := []*TreeNode{root.Left, root.Right}
		for len(stack) > 0 {
			children := []*TreeNode{}
			for len(stack) > 0 {
				n1, n2 := stack[0], stack[1]
				stack = stack[2:]
				if (n1 != nil && n2 == nil) || (n1 == nil && n2 != nil) {
					return false
				} else if n1 == nil && n2 == nil {
					continue
				} else if n1.Val != n2.Val {
					return false
				}
				children = append(children, n1.Left, n2.Right, n1.Right, n2.Left)
			}
			stack = children
		}
	}
	return true
}
